### ISSUE TITLE
splice junction loss

### ISSUE THREAD

**Issue by MiqG** [google-deepmind/alphagenome_research#24]
### Description

Hi,

thanks @tomwardio and @vdutor for solving my previous issue (https://github.com/google-deepmind/alphagenome_research/issues/22) so quickly!

As trying to fine-tune alphagenome on new splicing heads, I have noted that the splice junction loss from your repository differs from the splice junction loss explained in the paper's pseudocode:
- repo:
https://github.com/google-deepmind/alphagenome_research/blob/7046b72824b6e1914b012432e47fffb858e6c460/src/alphagenome_research/model/heads.py#L1191-L1246
https://github.com/google-deepmind/alphagenome_research/blob/7046b72824b6e1914b012432e47fffb858e6c460/src/alphagenome_research/model/losses.py#L45-L58
https://github.com/google-deepmind/alphagenome_research/blob/7046b72824b6e1914b012432e47fffb858e6c460/src/alphagenome_research/model/losses.py#L167-L184

- **pseudocode**:
```python
def soft_clip(x: Array) -> Array:
  return Where(x > 10.0, 2 * Sqrt(x * 10.0) - 10.0, x)

def multinomial_cross_entropy(
x: Array[D, A, N_tissues], targets: Array[D, A, N_tissues], axis: int
) -> Array[]:
  pred_ratios = (x + 1e-7) / (x + 1e-7).sum(axis=axis, keepdims=True)
  target_ratios = (targets + 1e-7) / (targets + 1e-7).sum(axis=axis, keepdims=True)
  return - (targets * Log(pred_ratios)).sum()

def poisson_loss(
x: Array[D, A, N_tissues], targets: Array[D, A, N_tissues], axis: int
) -> Array[]:
  sum_pred = x.sum(axis=axis)
  sum_targets = soft_clip(targets.sum(axis=axis))
  return (sum_pred - sum_targets * Log(sum_pred + 1e-7)).sum()
 
def junctions_loss(
x: Array[D, A, N_tissues], targets: Array[D, A, N_tissues]
) -> Array[]:
  ratios_loss = (multinomial_cross_entropy(x, targets, axis=0) + multinomial_cross_entropy(x, targets, axis=1))
  counts_loss = (poisson_loss(x, targets, axis=0) + poisson_loss(x, targets, axis=1))
  return 0.2 * ratios_loss + 0.04 * counts_loss
```

What loss exactly did you use to train AlphaGenome's splice junctions?

Here are some of the differences I could find and their impact:

- in the pseudocode's `multinomial_cross_entropy`, why is `target_ratios = (targets + 1e-7) / (targets + 1e-7).sum(axis=axis, keepdims=True)` computed if then it is not used in `- (targets * Log(pred_ratios)).sum()`?

- the repo `cross_entropy_loss`, returns a negative value when there isn't a splice junction. In all-zero-target cases, this can produce negative values when the mask contains valid entries, because the loss becomes essentially the log of the summed masked predictions. Is that expected, or should all-zero-target rows/columns be excluded from this ratio loss? This does not seem to be a problem in the pseudocode version.

- the pseudocode `poisson_loss` does not subtract the optimum/minimum value, so for well-fitted large counts it can become strongly negative. The repository implementation avoids this by subtracting `min_value`. Was this normalization used during training?

Here's a script to exemplify these issues on synthetic examples: 

[compare_junction_losses.py](https://github.com/user-attachments/files/27139348/compare_junction_losses.py)

Output:
```shell
============================================================
Case: All targets=0, valid mask, small pred (early training)
  pred shape: (8, 8, 2)
  target sum: 0.0000  pred sum: 1.2800
  JAX loss  : -5.0194
  Paper loss: +0.1024
  *** JAX loss is NEGATIVE ***

============================================================
Case: All targets=0, valid mask, pred=1.0
  pred shape: (8, 8, 2)
  target sum: 0.0000  pred sum: 128.0000
  JAX loss  : +7.3589
  Paper loss: +10.2400

============================================================
Case: Sparse targets, small pred
  pred shape: (8, 8, 2)
  target sum: 8.0000  pred sum: 1.2800
  JAX loss  : -3.2794
  Paper loss: +8.3731
  *** JAX loss is NEGATIVE ***

============================================================
Case: Well-fitted large counts (good model)
  pred shape: (8, 8, 2)
  target sum: 35.0000  pred sum: 35.0000
  JAX loss  : -24.5655
  Paper loss: -4.7222
  *** JAX loss is NEGATIVE ***
  *** Paper loss is NEGATIVE ***

============================================================
Case: D=A=256 (real padding size), all targets=0, small pred
  pred shape: (256, 256, 2)
  target sum: 0.0000  pred sum: 131.0720
  JAX loss  : -2.6226
  Paper loss: +10.4858
  *** JAX loss is NEGATIVE ***

============================================================
Case: D=A=256, sparse mask (10×10 valid), all targets=0
  pred shape: (256, 256, 2)
  target sum: 0.0000  pred sum: 131.0720
  JAX loss  : -9.2012
  Paper loss: +0.0160
  *** JAX loss is NEGATIVE ***
```

Thanks very much again! Best,

Miquel

### System info (python version, alphagenome_research version, etc.)

not applicable

**Comment by vdutor** [comment]
Thanks for spotting these discrepancies. To answer your questions:

(1) This is a typo in the paper pseudocode. The calculation should use target_ratios instead of raw targets.

(2) This was a bug in alphagenome_research due to missing epsilon smoothing on y_true. This is fixed now in https://github.com/google-deepmind/alphagenome_research/commit/de264f5f151b934c82f661e510b2409b84a40149 - thanks.

(3) Poisson loss without minimum subtraction. This was a simplification in the pseudocode. The Poisson loss in alphagenome_research, which subtracts min_value, is correct and was used during training. Note, however, that this constant offset (with respect to model parameters) should not affect gradients and therefore does not change training dynamics.

Also note that the paper pseudocode uses simple sums, while in practice we use means across (active) elements.  In general, I recommend following the code in alphagenome_research as it is closer to our internal implementation used for training than the paper pseudocode where we made certain simplifications for ease of understanding (e.g. missing masks).

**Comment by MiqG** [comment]
Nice! Thanks a lot again @vdutor and @tomwardio ! I will test these changes asap!

**Comment by MiqG** [comment]
Hi @vdutor and @tomwardio , I wanted to thank you for helping us! In the end, we managed to finetune on splicing tracks too: https://genomicsxai.github.io/blogs/2026-009/

Of course you are also in the acknowledgements of the blogpost!

Best,

Miquel
